Add torch::stable::Device

mikaylagawarecki opened the pull request:
Prior to this PR, the IValue <-> StableIValue conversion for `DeviceObjType` (aka c10::Device) was to pack it into the leading bits of the StableIValue (which is a uint64_t)

After this PR, the IValue <-> StableIValue conversion for `DeviceObjType` expects DeviceType to be packed into the upper 32 bits of StableIValue and DeviceIndex to be packed into the lower 32 bits

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #167362
* #166695
* #166694
* __->__ #166579



cc @jbschlosser

mikaylagawarecki committed:
Add torch::stable::Device

[ghstack-poisoned]

mikaylagawarecki commented on torch/csrc/stable/stableivalue_conversions.h:
```diff
@@ -128,6 +130,39 @@ struct FromImpl<ScalarType> {
+        return from(aoti_torch_device_type_mps());
+      case DeviceType::PrivateUse1:
+        return from(aoti_torch_device_type_privateuse1());
+      default:
```
I only added the ones for which there are aoti utils for now

mikaylagawarecki committed:
Update on "Add torch::stable::Device"

BC breaking note:

Prior to this PR, the IValue <-> StableIValue conversion for `DeviceObjType` (aka c10::Device) was to pack it into the leading bits of the StableIValue (which is a uint64_t)

After this PR, the IValue <-> StableIValue conversion for `DeviceObjType` expect a pointer to a DeviceHandle to be the StableIValue representation. c10::Device is represented as a torch::stable::Device in user extensions.





cc jbschlosser

[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Add torch::stable::Device"

BC breaking note:

Prior to this PR, the IValue <-> StableIValue conversion for `DeviceObjType` (aka c10::Device) was to pack it into the leading bits of the StableIValue (which is a uint64_t)

After this PR, the IValue <-> StableIValue conversion for `DeviceObjType` expect a pointer to a DeviceHandle to be the StableIValue representation. c10::Device is represented as a torch::stable::Device in user extensions.





cc jbschlosser

[ghstack-poisoned]

janeyx99 commented on test/cpp_extensions/libtorch_agnostic_extension/libtorch_agnostic/csrc/kernel.cpp:
```diff
@@ -486,6 +487,164 @@ STABLE_TORCH_LIBRARY_IMPL(libtorch_agnostic, CompositeExplicitAutograd, m) {
+  m.def("test_device_set_index() -> int");
+  m.def("test_device_has_index() -> bool");
+  m.def("test_device_type() -> bool");
+  m.def("test_device_argument(Device device) -> bool");
```
A more highlevel comment: I'm thinking we can leverage the functions in this file better + test more flavored use cases by having a function that just calls the APIs (e.g., has_index, set_index, device_equals, each constructor). This way, we test more in the python world without bloating kernel.cpp.

janeyx99 commented on torch/csrc/shim_common.cpp:
```diff
@@ -225,6 +230,65 @@ AOTI_TORCH_EXPORT AOTITorchError torch_library_impl(
+    int32_t device_type,
+    int32_t device_index,
+    DeviceHandle* ret_device) {
+  AOTI_TORCH_CONVERT_EXCEPTION_TO_ERROR_CODE({
```
We should eventually wean off the AOTI macros too...not within scope for this PR, but just flagging as another thing we should do

janeyx99 commented on torch/csrc/stable/device_struct.h:
```diff
@@ -0,0 +1,118 @@
+  /* implicit */ Device(DeviceType type, int32_t index = -1) {
+    DeviceHandle ret;
+    TORCH_ERROR_CODE_CHECK(
+        torch_create_device(static_cast<int32_t>(type), index, &ret));
```
I suspect we need to call `from` and `to` on the type here instead of static casting -- and everywhere else we input the type into the shim

mikaylagawarecki committed:
Update on "Add torch::stable::Device"

BC breaking note:

Prior to this PR, the IValue <-> StableIValue conversion for `DeviceObjType` (aka c10::Device) was to pack it into the leading bits of the StableIValue (which is a uint64_t)

After this PR, the IValue <-> StableIValue conversion for `DeviceObjType` expect a pointer to a DeviceHandle to be the StableIValue representation. c10::Device is represented as a torch::stable::Device in user extensions that use `STABLE_TORCH_LIBRARY`.





cc jbschlosser

[ghstack-poisoned]

janeyx99 commented on torch/csrc/shim_common.cpp:
```diff
@@ -109,8 +110,12 @@ static c10::IValue to_ivalue(
     case c10::TypeKind::DeviceObjType: {
-      return c10::IValue(torch::stable::detail::_to<c10::Device>(
-          stable_ivalue, extension_build_version));
+      auto deleter = [](DeviceOpaque* device) { torch_delete_device(device); };
```
I need to confirm the idea I have works (I'm going to do it for StableList), but in essence it would be to make a new reference from the devicehandle, and then explicitly delete the old devicehandle. No need to change code here until I confirm it works.

mikaylagawarecki committed:
Update on "Add torch::stable::Device"

BC breaking note:

Prior to this PR, the IValue <-> StableIValue conversion for `DeviceObjType` (aka c10::Device) was to pack it into the leading bits of the StableIValue (which is a uint64_t)

After this PR, the IValue <-> StableIValue conversion for `DeviceObjType` expect a pointer to a DeviceHandle to be the StableIValue representation. c10::Device is represented as a torch::stable::Device in user extensions that use `STABLE_TORCH_LIBRARY`.





cc jbschlosser

[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Add torch::stable::Device"

BC breaking note:

Prior to this PR, the IValue <-> StableIValue conversion for `DeviceObjType` (aka c10::Device) was to pack it into the leading bits of the StableIValue (which is a uint64_t)

After this PR, the IValue <-> StableIValue conversion for `DeviceObjType` expect a pointer to a DeviceHandle to be the StableIValue representation. c10::Device is represented as a torch::stable::Device in user extensions that use `STABLE_TORCH_LIBRARY`.





cc jbschlosser

[ghstack-poisoned]

mikaylagawarecki commented on torch/csrc/stable/stableivalue_conversions.h:
```diff
@@ -200,6 +235,30 @@ struct FromImpl<torch::stable::Tensor> {
+// Specialization for torch::stable::Device => StableIValue
+// Pack the device type and index into a StableIValue in a platform-independent
+// format. We use the shim representation for DeviceType (int32_t) for ABI
+// stability. StableIValue layout: DeviceType (shim int32_t) in lower 32 bits,
```
I didn't use the default `from` because I worry that struct padding might differ across systems

I reversed the order of DeviceType and DeviceIndex because DeviceIndex can be negative (putting it in front gives it the sign bit)

janeyx99 commented on torch/csrc/shim_common.cpp:
```diff
@@ -110,12 +115,24 @@ static c10::IValue to_ivalue(
+          static_cast<int8_t>(stable_ivalue & 0xFF));
+      int32_t device_index =
+          static_cast<int32_t>((stable_ivalue >> 32) & 0xFFFFFFFF);
+      TORCH_CHECK(
```
Why do we want this check? Does this mean that we need to remove this check once we support more than 128 GPUs?

janeyx99 commented on torch/csrc/stable/device.h:
```diff
@@ -1,4 +1,231 @@
+// It has some slight modifications e.g. TORCH_INTERNAL_ASSERT_DEBUG_ONLY ->
+// STD_TORCH_CHECK We chose to copy it rather than moving it to headeronly as
+// (1) Device is < 8 bytes so the *Handle approach used for tensor doesn't make
+// sense (2) c10::Device is not header-only due to its string constructor
```
```suggestion
// sense (2) c10::Device is not header-only due to its string constructor.
```

mikaylagawarecki committed:
Update on "Add torch::stable::Device"

Prior to this PR, the IValue <-> StableIValue conversion for `DeviceObjType` (aka c10::Device) was to pack it into the leading bits of the StableIValue (which is a uint64_t)

After this PR, the IValue <-> StableIValue conversion for `DeviceObjType` expects DeviceIndex to be packed into the upper 32 bits and DeviceType to be packed into the lower 32 bits of the StableIValue





cc jbschlosser

[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Add torch::stable::Device"

Prior to this PR, the IValue <-> StableIValue conversion for `DeviceObjType` (aka c10::Device) was to pack it into the leading bits of the StableIValue (which is a uint64_t)

After this PR, the IValue <-> StableIValue conversion for `DeviceObjType` expects DeviceIndex to be packed into the upper 32 bits and DeviceType to be packed into the lower 32 bits of the StableIValue





cc jbschlosser

[ghstack-poisoned]

janeyx99 reviewed (approved):
yay!

mikaylagawarecki committed:
Update on "Add torch::stable::Device"

Prior to this PR, the IValue <-> StableIValue conversion for `DeviceObjType` (aka c10::Device) was to pack it into the leading bits of the StableIValue (which is a uint64_t)

After this PR, the IValue <-> StableIValue conversion for `DeviceObjType` expects DeviceType to be packed into the upper 32 bits of StableIValue and DeviceIndex to be packed into the lower 32 bits




cc jbschlosser

[ghstack-poisoned]
